native/src/main/java/io/ayash/blockingqueue/BlockingQueue.java

```diff
@@ -0,0 +1,165 @@
+    private int size = 0;
+
+    public BlockingQueue(int capacity) {
+        this.capacity = capacity;
+        queue = new Object[capacity];
+    }
+
+    public void put(Object item) throws InterruptedException {
+        Objects.requireNonNull(item);
+        lock.lockInterruptibly();
+        try {
+            while (size == capacity) {
+                notFull.await();
+            }
+            queue[tail] = item;
+            tail = (tail + 1) % capacity;
+            size++;
+            notEmpty.signal();
+        } finally {
+            lock.unlock();
+        }
+    }
+
+
+    public Object take() throws InterruptedException {
+        lock.lockInterruptibly();
+        try {
+            while (size == 0) {
+                notEmpty.await();
+            }
+            Object item = queue[head];
+            head = (head + 1) % capacity;
+            size--;
+            notFull.signal();
+            return item;
+        } finally {
+            lock.unlock();
+        }
+    }
+
+    public boolean add(Object item) {
+        Objects.requireNonNull(item);
+        lock.lock();
+        try {
+            if (size == capacity) {
+                throw new IllegalStateException("Queue is full");
+            }
+            queue[tail] = item;
+            tail = (tail + 1) % capacity;
+            size++;
+            notEmpty.signal();
+            return true;
+        } finally {
+            lock.unlock();
+        }
+    }
+
+    public boolean offer(Object item) {
+        Objects.requireNonNull(item);
+        lock.lock();
+        try {
+            if (size == capacity) {
+                return false;
+            }
+            queue[tail] = item;
+            tail = (tail + 1) % capacity;
+            size++;
+            notEmpty.signal();
+            return true;
+        } finally {
+            lock.unlock();
+        }
+    }
+
+    public boolean offer(Object item, long timeout) throws InterruptedException {
+        Objects.requireNonNull(item);
+        lock.lockInterruptibly();
+        try {
+            long nanos = timeout * 1_000_000;
+            while (size == capacity) {
+                if (nanos <= 0) {
+                    return false;
+                }
+                nanos = notFull.awaitNanos(nanos);
+            }
+            queue[tail] = item;
+            tail = (tail + 1) % capacity;
+            size++;
+            notEmpty.signal();
+            return true;
+        } finally {
+            lock.unlock();
+        }
+    }
+
+
+    public boolean contains(Object item) {
+        Objects.requireNonNull(item);
+        lock.lock();
+        try {
+            for (int i = 0; i < size; i++) {
+                if (queue[(head + i) % capacity].equals(item)) {
+                    return true;
+                }
+            }
+            return false;
+        } finally {
+            lock.unlock();
+        }
+    }
+
+    public Object peek() {
+        lock.lock();
+        try {
```

Comment:
Instead, You can replace `if (size == 0) return null` in `peek()` with a ternary operator for brevity.

```java 
    public Object peek() {
        lock.lock();
        try {
            return size == 0 ? null : queue[head];
        } finally {
            lock.unlock();
        }
    }
```